pandas/core/series.py

```diff
@@ -623,6 +623,29 @@ def view(self, dtype=None):
         return self._constructor(self._values.view(dtype),
                                  index=self.index).__finalize__(self)
 
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
```

Comment:
Normally we recommend doing a type check and returning `NotImplemented` if you see an unknown type . If `__array_ufunc__` always tries to do the operation, you can get non-commutative operations, e.g., `a + b != b + a`.

For example, consider another library like xarray that implements its own version of a `pandas.Series`. `Series + DataArray` and `DataArray + Series` should consistently give either `Series`, `DataArray` or an error (probably the safest choice) -- they shouldn't just return a result of the same type as the first argument.

This might even be an issue for `Series + DataFrame` or `np.add(Series, DataFrame)` -- you probably want to delegate to the DataFrame in that case, not the Series.

Reply:
I was thinking somehow: the `__array_ufunc__` of the underlying values we call here is already doing the check, so I don't need to do it here again. 
But indeed, that will probably mess up such cases. 

The problem is that for now the `__array_ufunc__` uses our own implementations of all arithmetic/comparison operators, so I should somehow list here everything that all those different ops accept, which might be a bit annoying.

Reply:
I doubt you need to list types dependent on the operators -- it's *probably* enough to list the types that can be used in valid operations with `pandas.Series`.

It's true that you would possibly need to recurse into extension arrays to figure this out all the valid types, and it may not be easy to enumerate all scalar types, e.g., if IPAddressExtensionArray can handle arithmetic with IPAddress, how could pandas.Series know about that?

Reply:
Yes, I think it are mainly the scalar values which might be hard to list (apart from those, we would probably accept Series, Index, array, and a bunch of array-likes (list, tuple, ..) for the Series case). 
But eg a Series with datetimelike values can already handle our own Timestamp, Timedelta and offset objects, numpy scalars, `datetime` standard library scalars, integer, ..), and as I understand those should all be listed here? 
(but anyhow, if we implement an `__array_ufunc__` for eg PeriodArray, we would need to do that there as well, so need to make this list in any case)

But what you mention about custom ExtensionArrays is then even another problem, that we don't know what they would accept. By "recurse into extension arrays", would you mean something like checking the `ExtensionArray._HANDLED_TYPES` (and then expecting EA authors to set this properly)? Or some other way to inspect?

Reply:
> By "recurse into extension arrays", would you mean something like checking the `ExtensionArray._HANDLED_TYPES` (and then expecting EA authors to set this specific property)? Or some other way to inspect?

Yeah, that's about the best I can imagine. But I agree, it doesn't seem easy! There's no general way to detect what types another object's `__array_ufunc__` method can handle without actually calling it.

It is probably reasonable to trade-off correctness for generality here, but we should be cognizant that we are making this choice. Alternatively, we might maintain an explicit "blacklist" of types which pandas can't handle, though this gets tricky with non-required imports (e.g., `dask` and `xarray` objects).

Reply:
+1 to instructing (requiring?) ExtensionArray authors to implement `_HANDLED_TYPES` if they implement `__array_ufunc__`.

Reply:
@shoyer in light of the above, how would you define the handled types for an object dtyped Series, in which you in principle can put any python object you want? For numpy arrays, things like addition, work element-wise for object dtype, so the following works:

```
In [101]: class A():
     ...:     def __init__(self, x):
     ...:         self.x = x
     ...:     def __add__(self, other):
     ...:         return A(self.x + other.x)
     ...:     def __repr__(self):
     ...:         return "A(x={})".format(self.x)
     ...:          

In [102]: a = np.array([A(1), A(2)], dtype=object)

In [103]: np.add(a, A(3))
Out[103]: array([A(x=4), A(x=5)], dtype=object)

In [104]: np.add(pd.Series(a), A(3))
Out[104]: 
0    A(x=4)
1    A(x=5)
dtype: object
```

Doing a more strict check for allowed types, then the above will no longer work.

Reply:
How is numpy actually dealing with the above? It makes an exception for object dtype?

So two ideas that might be useful here:

* We can make an exception for object dtype, and in that case allow more unknown objects
* We could check for objects that implement `__array_ufunc__` themselves? If they have the attribute, and are not known to us (not in our `_HANDLED_TYPES` list, eg dask or xarray objects), we raise NotImplemented, otherwise we pass through. 
  That at least solves the problem for the more known objects that will probably have the interface, but of course not yet for custom container types that do not implement it.

Reply:
These are good questions. In principle NumPy follows the rules described at http://www.numpy.org/neps/nep-0013-ufunc-overrides.html#behavior-in-combination-with-python-s-binary-operations but I think there is something missing for unknown scalars / object arrays...

Reply:
See https://github.com/numpy/numpy/issues/12258 for a description of how NumPy does things.

Reply:
To make sure I understand this thread: @shoyer kicked it off by saying

> Normally we recommend doing a type check and returning NotImplemented if you see an unknown type

We can handle the scalar types of EAs by requiring that they implement a `_HANDLED_TYPEs`, which would include their scalar type. Then if all the (unboxed from Series / Index) inputs are in handled types, we proceed.

object-dtype complicates things. Did we decide what to do there? Skip the check?